src/model/introspect.rs

```diff
@@ -143,32 +149,29 @@ impl Introspector {
         self.pages > 0
     }
 
-    /// Query for all matching elements.
-    pub fn query(&self, selector: Selector) -> Vec<Content> {
-        self.all().filter(|elem| selector.matches(elem)).cloned().collect()
+    /// Gets the element at the given index.
+    pub fn get(&self, index: usize) -> Option<&Content> {
+        self.elems.get(index).map(|(elem, _)| elem)
     }
 
-    /// Query for all matching element up to the given location.
-    pub fn query_before(&self, selector: Selector, location: Location) -> Vec<Content> {
-        let mut matches = vec![];
-        for elem in self.all() {
-            if selector.matches(elem) {
-                matches.push(elem.clone());
-            }
-            if elem.location() == Some(location) {
-                break;
-            }
-        }
-        matches
+    /// Collects the element in a vec
+    pub fn to_vec(&self) -> Vec<&Content> {
+        self.elems.iter().map(|(elem, _)| elem).collect()
+    }
+
+    /// Get an element from the position cache.
+    pub fn position_cache(&self, location: &Location) -> Option<usize> {
+        self.position_cache.get(location).copied()
+    }
```

Comment:
only methods that are actually used externally on a `Tracked<Introspector` should be in the `#[track]` impl.